src/environments.js

```diff
@@ -1,18 +1,96 @@
-/* eslint no-constant-condition:0 */
+// @flow
 import ParseNode from "./ParseNode";
 import ParseError from "./ParseError";
 
+import type Parser from "./Parser";
+import type {ArgType, Mode, StyleStr} from "./types";
+
+/**
+ * The context contains the following properties:
+ *  - mode: current parsing mode.
+ *  - envName: the name of the environment, one of the listed names.
+ *  - parser: the parser object.
+ *  - positions: the positions associated with these arguments from args.
+ */
+type EnvContext = {
+  mode: Mode,
+  envName: string,
+  parser: Parser,
+  positions: number[],
+};
+
+/**
+ * List of ParseNodes followed by an array of positions.
+ * Returned by `Parser`'s `parseArguments`.
+ */
+type ParseResult = (ParseNode | number[])[] | ParseNode;
+
+/**
+ * The handler function receives two arguments
+ *  - context: information and references provided by the parser
+ *  - args: an array of arguments passed to \begin{name}
+ */
+type EnvHandler = (context: EnvContext, args: ParseNode[]) => ParseResult;
+
+/**
+ *  - numArgs: (default 0) The number of arguments after the \begin{name} function.
+ *  - argTypes: (optional) Just like for a function
+ *  - allowedInText: (default false) Whether or not the environment is allowed
+ *  i                nside text mode (not enforced yet).
+ *  - numOptionalArgs: (default 0) Just like for a function
+ */
+type EnvProps = {
+    numArgs?: number,
+    argTypes?: ArgType[],
+    allowedInText?: boolean,
+    numOptionalArgs?: number,
+};
+
+type EnvData = {
+    numArgs: number,
+    argTypes?: ArgType[],
+    greediness: number,
+    allowedInText: boolean,
+    numOptionalArgs: number,
+    handler: EnvHandler,
+};
+const environments: {[string]: EnvData} = {};
+export default environments;
+
+// Data stored in the ParseNode associated with the environment.
+type ArrayColSpec = { type: "separator", separator: string } | {
+    type: "align",
+    align: string,
+    pregap?: number,
+    postgap?: number,
+};
+type EnvNodeData = {
```

Comment:
`ArrayEnvNode` or `ArrayNode` would be better as this type is specific to arrays.  For consistency `ArrayColSpec` could be `AlignNode` and we could even introduce a `SeparateNode` type.

Reply:
`ArrayEnvNode` is more descriptive. So I'm using that.
Is `AlignNode` the most appropriate name? It's not a `ParseNode`.

Reply:
Also, per `ArrayEnvNode`, it's not a `ParseNode`, but the data stored in the `ParseNode`. Would you still prefer to call it `ArrayEnvNode`? Or would `ArrayEnvNodeData` (a little long, I know) be more appropriate to avoid confusion?

Reply:
In the currently-pushed version, I went with `AlignSpec` and `ArrayEnvNodeData`. I'm very open to changing per your feedback.

Reply:
I don't see a need to differentiate parse nodes from environment nodes as they end up being part of the same tree at the end of the day.  I think eventually we'll types for each of the different nodes the parser can generate and eventually `ParseNode` will become:
```
ParseNode = FracNode | ExpNode | ... | ArrayNode | AlignNode | ... 
```
Environments are just one type of thing that the parser parses as it does its work, it just happens that the environment parsing code is in a separate file.

Reply:
I'm sorry, I didn't quite understand this comment and the future evolution you intend. What I meant was: for `result: ArrayEnvNodeData`, we have
```javascript
    return new ParseNode(result.type, result, parser.mode);
```
So the `ParseNode` instance contains a _payload_ that is of type `ArrayEnvNodeData` struct, which is why I was thinking about the `Data` suffix.

So did you mean that the payload would become a union type (e.g. `ParseNodeData` or something briefer)?
I'm guessing you didn't mean that the `ParseNode` class would be dissolved in favor of a union type (given that we still want to encapsulate the common tokenizer metadata for error reporting)?

Reply:
I was thinking that eventually we'd have a union type to specify the different types in the tree.  It's possible that `ParseNode` wraps these data nodes and adds additional data like error reporting.  I have spent any time thinking trough the details.  The high-level goal is for nodes with in the parse tree to be more regular and for all them to be statically typed.

Reply:
Ok, so what you're saying is that the _payload_ of `ParseNode` is still considered an undecorated "Node". And hence we should use the `Node` suffix rather than `NodeData`?
Is that your preference?

Reply:
Ah, I see that you merged it. Please LMK what nomenclature you had in mind. I'm happy to update these as we continue in the future.